Add Initial Tests for a Model, Controller and Module

SanGeeky opened the pull request:
- Add Test for Article Model, with three contexts, database, model, crud.
- Add Test For Articles Controller, with users authenticated and no authenticated
- Add Test for Application Helper Module.

SanGeeky committed:
Add Test with RSpec

- Install gem rails-rspec and rexml
- Add generator to associate future models to rspec
- Add Test for Article Model and Controller, Test for Application
Helper

SanGeeky committed:
Add unit tests for a Model, Controller and Module

- Additional gem, shoulda-matchers
- Add Test for Article Model, with three contexts
- Add Test For Articles Controller, with users authenticated and no authenticated
- Add Test for Application Helper Module.

mauriciosabogal commented on Gemfile:
```diff
@@ -28,6 +28,9 @@ gem 'bcrypt', '~> 3.1.7'
 gem 'bootsnap', '>= 1.4.4', require: false
 
+# Features an easy-to-use API
+gem 'rexml', '~> 3.2.4'
```
Why do you need this gem?

SanGeeky replied:
Because i am getting this error when the tests are loaded:
```
Failure/Error: require File.expand_path('../config/environment', __dir__)

LoadError:
  cannot load such file -- rexml/document

```
But now I moved to test group

mauriciosabogal commented on Gemfile:
```diff
@@ -50,6 +53,16 @@ group :test do
+  # Add RSpec testing library
+  gem 'rspec-rails'
+  # Test templates
+  gem 'rails-controller-testing'
```
Why do you need this gem `rails-controller-testing`?

SanGeeky replied:
Is used for the matcher `render_template`, and i was getting this error:
```
NoMethodError:
       assert_template has been extracted to a gem. To continue using it,
               add `gem 'rails-controller-testing'` to your Gemfile.
```
In Rails 5 the method was removed

mauriciosabogal commented on spec/models/articles_spec.rb:
```diff
@@ -0,0 +1,81 @@
+
+  context 'database' do
+    describe 'columns' do
+      it { should have_db_column(:title) }
```
Use `is_expected` syntax instead of `should`. `should` might generate some issues.

SanGeeky replied:
Did it !

mauriciosabogal commented on spec/models/articles_spec.rb:
```diff
@@ -0,0 +1,81 @@
+    end
+  end
+
+  context 'crud' do
```
There is no need of testing these crud methods because they are part of Rails. You only have to test your code.

SanGeeky replied:
Thanks for ther advise

mauriciosabogal commented on spec/helpers/application_helper_spec.rb:
```diff
@@ -0,0 +1,39 @@
+RSpec.describe ApplicationHelper, type: :helper do
+  let(:user) { create :user }
+
+  context "user unauthenticated" do
```
Each method should have its own context.
```
context '#current_user' do
  ...
end

context '#user_logged_in?' do
  ..
end
```

Also, take a look what's the correct way for context descriptions.

`it 'current ser' do` should be `it 'returns the current user' do`

SanGeeky replied:
```ruby
  context '#current_user' do
    it 'returns the current ser' do
      session[:user_id] = user.id
      expect(current_user).to eq(user)
    end

    it 'not returns the current user' do
      expect(current_user).to be_nil
    end
  end

  context '#user_logged_in?' do
    it 'user logged in' do
      session[:user_id] = user.id
      expect(user_logged_in?).to eq(user)
    end

    it 'user not logged in' do
      expect(user_logged_in?).to be_nil
    end
  end
```

Thakns for the advise

mauriciosabogal commented on spec/requests/articles_controller_spec.rb:
```diff
@@ -0,0 +1,133 @@
+RSpec.describe ArticlesController, type: :controller do
+  context 'articles user not authenticated' do
+    describe '#index' do
+      before { get :index }
```
This should be the `subject`.

SanGeeky replied:
Done !

mauriciosabogal commented on spec/requests/articles_controller_spec.rb:
```diff
@@ -0,0 +1,133 @@
+    end
+
+    describe '#show' do
+      subject { create :article }
```
This should be a `before` block. The subject should be `get :index`.

SanGeeky replied:
Done!

mauriciosabogal commented on spec/requests/articles_controller_spec.rb:
```diff
@@ -0,0 +1,133 @@
+    describe '#update' do
+      let(:article) { create :article }
+
+      before { get :create, params: { id: article.id } }
```
`get :create` is not the right action here.

SanGeeky replied:
I update to `put` in this case.

mauriciosabogal commented on spec/requests/articles_controller_spec.rb:
```diff
@@ -0,0 +1,133 @@
+    describe '#destroy' do
+      let(:article) { create :article }
+
+      before { get :destroy, params: { id: article.id } }
```
`get :destroy` is not the right action here.

SanGeeky replied:
I update to `delete` in this case.

mauriciosabogal commented on spec/requests/articles_controller_spec.rb:
```diff
@@ -0,0 +1,133 @@
+  end
+end
+
+RSpec.describe ArticlesController, type: :request do
```
Why are you testing the actions twice? You have `controller` and `request` tests.

SanGeeky replied:
I testing the Controller as `controller` type, and `request` type because with the `request` type I can log in with the user. Should I use only one type ?

mauriciosabogal replied:
Yes, please use `request` type.

SanGeeky committed:
Refactor Initial Tests
- Articles Controller using :request type
- Remove context crud form Article Model
- Reformat Application Helper
- rexml gem in test group
- Replace should keyword by is_expected
